python/ambassador/ir/irhealthchecks.py

```diff
@@ -0,0 +1,235 @@
+                            log_level=logging.ERROR,
+                        )
+                    addHeaders = self.generate_headers_to_add(request_headers_to_add)
+                    if len(addHeaders) > 0:
+                        http_mapper["request_headers_to_add"] = addHeaders
+                request_headers_to_remove = http_health_check.get("remove_request_headers", None)
+                if request_headers_to_remove is not None:
+                    if not isinstance(request_headers_to_remove, list):
+                        self.post_error(
+                            f"IRHealthChecks: remove_request_headers must be a list. Ignoring field for health-check: {hc}",
+                            log_level=logging.ERROR,
+                        )
+                    else:
+                        http_mapper["request_headers_to_remove"] = request_headers_to_remove
+
+                host = http_health_check.get("hostname", None)
+                if host is not None:
+                    http_mapper["host"] = host
+
+                # Process the expected statuses
+                expected_statuses = http_health_check.get("expected_statuses", None)
+                if expected_statuses is not None:
+                    validStatuses = []
+                    for statusRange in expected_statuses:
+                        startCode = int(statusRange["start"])
+                        endCode = int(statusRange["end"])
+                        if startCode < 100 or startCode >= 600:
+                            self.post_error(
+                                f"IRHealthChecks: expected_statuses: {startCode} must be an integer >= 100 and < 600. Ignoring expected status for health-check {hc}",
+                                log_level=logging.ERROR,
+                            )
+                            continue
+                        if endCode < 100 or endCode >= 600:
+                            self.post_error(
+                                f"IRHealthChecks: expected_statuses: {endCode} must be an integer >= 100 and < 600. Ignoring expected status for health-check {hc}",
+                                log_level=logging.ERROR,
+                            )
+                            continue
+                        if startCode > endCode:
+                            self.post_error(
+                                f"IRHealthChecks: expected_statuses: status range start value {startCode} cannot be higher than the end {endCode} for range. Ignoring expected status for health-check {hc}",
+                                log_level=logging.ERROR,
+                            )
+                            continue
+
+                        # We add one to the end code because by default Envoy expects the start of the range to be
+                        # inclusive, but the end of the range to be exclusive. Lets just make both inclusive for simplicity.
```

Comment:
Is it better to diverge from Envoy in this case? I mean most people follow our docs which we can call this out but for folks that read envoy docs this would be confusing...no hard opinion on this one but not sure it is a benefit to diverge...

Reply:
I'm of the opinion that if you mark something as the end/start it should be included. I was trying to make it easier to understand for our users rather than cater to people who might know what the underlying envoy config is. To me ranges where the start is inclusive and the end is exclusive always irritated me, it might make sense when writing code but I think it makes a lot less sense when specifying configuration. 

I'm open to changing this though if there is benefit to doing so.

Reply:
What if a user sets the end code to 599 which passes validation and then the plus 1 logic here would set the end to 600 correct?

If I read the docs it doesn't allow for the value to be 600 so would this throw invalid envoy configuration?

https://www.envoyproxy.io/docs/envoy/v1.23.0/api-v3/config/core/v3/health_check.proto.html?highlight=expected_statuses#config-core-v3-healthcheck-httphealthcheck

The plus one logic should probably move above the range error checks? Can we have a test that verifies the ends to ensure nothing causes issues.

Reply:
Great suggestion!

Reply:
> What if a user sets the end code to 599 which passes validation and then the plus 1 logic here would set the end to 600 correct?
If I read the docs it doesn't allow for the value to be 600 so would this throw invalid envoy configuration?

nope, the following Envoy config example is perfectly legal and works fine because Envoy is treating `end:  600` as exclusive. `start` can never be 600 because of CRD validation but the IR check for validity on these values needs fixed:
```
            "http_health_check": {
              "expected_statuses": [
                {
                  "end": 600,
                  "start": 200
                }
              ],
              "path": "/health"
            },
````